packages/lib/EventDispatcher.ts

```diff
@@ -0,0 +1,71 @@
+type Listener<Value> = (data: Value)=> void;
+
+/**
+ * Sends notifications/allows listening for events.
+ *
+ * @param EventKeyType Possible identifiers for events.
+ * @param EventType Type of data associated with events.
+ */
+export default class EventDispatcher<EventKeyType extends string|symbol|number, EventType> {
+	// Partial marks all fields as optional. To initialize with an empty object, this is required.
+	// See https://stackoverflow.com/a/64526384
+	private listeners: Partial<Record<EventKeyType, Array<Listener<EventType>>>>;
+	public constructor() {
+		this.listeners = {};
+	}
+
+	/**
+	 * Sends a notification to all listeners subscribed to the given [notificationKind].
+	 * @param eventType Type of the notification.
+	 * @param value Data associated with the notification.
+	 */
+	dispatch(eventName: EventKeyType, event: EventType = null) {
+		if (!this.listeners[eventName]) return;
+
+		const ls = this.listeners[eventName];
+		for (let i = 0; i < ls.length; i++) {
+			ls[i](event);
+		}
+	}
+
+	/**
+	 * Listen for a specific type of event.
+	 *
+	 * @param eventType Type of event that causes `callback` to be called.
+	 * @param callback Callback called when an event of `eventType` is fired.
+	 * @returns An object with a `remove` function. Calling `remove` unregisters the listener.
+	 */
+	on(eventName: EventKeyType, callback: (data: EventType)=> void) {
```

Comment:
No inline definition please, please create a new type CallbackHandler for example. And make sure to update the off call too.